lib/ClangToAst.ml

```diff
@@ -1041,41 +1053,62 @@ let rec translate_expr (env : env) ?(must_return_value=false) (e : Clang.Ast.exp
         in
         let base = translate_expr env base in
 
+        let lid =
+          Helpers.assert_tlid
+            (if arrow then
+               Helpers.assert_tbuf base.typ
+             else
+               base.typ)
+        in
+
         let f =
           match field with
           | FieldName { desc; _ } -> get_id_name desc.name
-          | _ -> failwith "member node: only field accesses supported"
+        | _ -> failwith "member node: only field accesses supported"
         in
 
-        let field_t =
-          let lid =
-            Helpers.assert_tlid
-              (if arrow then
-                 Helpers.assert_tbuf base.typ
-               else
-                 base.typ)
-          in
-          match LidMap.find_opt lid !type_def_map with
+        begin match LidMap.find_opt lid !type_def_map with
+          | Some (lazy (Variant branches)) ->
+              let branch = match List.find_opt (fun b -> fst b = f) branches with
+                | Some b -> b
+                | None -> fatal_error "Field %s of %a not found in tagged union" f plid lid
+              in
+
+              begin
+              match snd branch with
+              | [_] ->
+                  let var = match base.node with | EBound n -> List.nth env.vars n | _ -> failwith "Tagged union access is only supported on a variable" in
+                  begin match !(fth4 var) with
+                  | Some { case; var } when case = f ->
+                      let e, _, _ = find_var env var in
+                      e
+                  | _ -> failwith "Tagged union variable is not in the correct case"
+                  end
+              | _ -> failwith "More than one field in tagged union case"
+              end
           | Some (lazy (Flat fields)) ->
-              if List.mem_assoc (Some f) fields then
-                fst (List.assoc (Some f) fields)
-              else
-                fatal_error "Field %s of %a not found in struct def (available fields are: %s)"
-                  f plid lid
-                  (String.concat ", " (List.map (fun (f, _) -> Option.value ~default:"<noname>" f) fields))
-          | Some _ -> fatal_error "Taking a field of %a which is not a struct" plid lid
-          | None -> fatal_error "Taking a field of %a which is not in the map" plid lid
-        in
 
-        if not arrow then
-          (* base.f *)
-          with_type field_t (EField (base, f))
-        else
-          (* base->f *)
-          let deref_base =
-            Helpers.(with_type (assert_tbuf base.typ) (EBufRead (base, Helpers.zero_usize)))
-          in
-          with_type field_t (EField (deref_base, f))
+              let field_t =
```

Comment:
introduce a helper for this whole chunk of code called lookup_field?

Reply:
basically starting at begin match LidMap.find_opt lid !type_def_map with

Reply:
This code actually does a bit more than just looking up the field, it constructs the Ast expression corresponding to the field access. In the case we are accessing a structure, this is nicely split between looking up the corresponding field, and then building an EField expression (with possibly a deref in case of an arrow).
The tagged union case however does not create an EField, but simply returns the variable (as an expr) corresponding to the current variant.

I'm however happy to extract all of the pattern-matching into an auxiliary `construct_field_expression` (or any better name) if you think this is useful.